sdk/communication/azure-communication-callautomation/azure/communication/callautomation/_shared/token_exchange.py

```diff
@@ -0,0 +1,170 @@
+
+
+TEAMS_EXTENSION_SCOPE_PREFIX = "https://auth.msft.communication.azure.com/"
+COMMUNICATION_CLIENTS_SCOPE_PREFIX = "https://communication.azure.com/clients/"
+TEAMS_EXTENSION_ENDPOINT = "/access/teamsPhone/:exchangeAccessToken"
+TEAMS_EXTENSION_API_VERSION = "2025-03-02-preview"
+COMMUNICATION_CLIENTS_ENDPOINT = "/access/entra/:exchangeAccessToken"
+COMMUNICATION_CLIENTS_API_VERSION = "2024-04-01-preview"
+
+class TokenExchangeClient:
+    """Represents a client that exchanges an Entra token for an Azure Communication Services (ACS) token."""
+
+    def __init__(self, resource_endpoint: str, token_credential: TokenCredential, scopes: Optional[List[str]] = None, pipeline_transport: Any = None):
+        self._resource_endpoint = resource_endpoint
+        self._scopes = scopes or ["https://communication.azure.com/clients/.default"]
+        self._token_credential = token_credential
+        self._pipeline = self._create_pipeline_from_options(token_credential, scopes, pipeline_transport)
+
+    def _create_pipeline_from_options(self, token_credential, scopes, pipeline_transport):
+        auth_policy = BearerTokenCredentialPolicy(token_credential, scopes)
+        entra_token_guard_policy = EntraTokenGuardPolicy()
+        policies = [auth_policy, entra_token_guard_policy]
+        if pipeline_transport:
+            return Pipeline(policies=policies, transport=pipeline_transport)
+        from azure.core.pipeline.transport import RequestsTransport
+        return Pipeline(policies=policies, transport=RequestsTransport())
+
+    def exchange_entra_token(self) -> AccessToken:
+        message = _TokenExchangeUtils.create_request_message(self._resource_endpoint, self._scopes)
+        response = self._pipeline.run(message)
+        return self._parse_access_token_from_response(response)
+
+    def _parse_access_token_from_response(self, response: PipelineResponse) -> AccessToken:
+        if response.http_response.status_code == 200:
+            try:
+                content = response.http_response.text()
+                data = json.loads(content)
+                access_token_json = data["accessToken"]
+                token = access_token_json["token"]
+                expires_on = access_token_json["expiresOn"]
+                expires_on_epoch = _TokenExchangeUtils.parse_expires_on(expires_on, response)
+                if expires_on_epoch is None:
+                    raise ClientAuthenticationError("Failed to parse 'expiresOn' value from access token response")
+                return AccessToken(token, expires_on_epoch)
+            except Exception as ex:
+                raise ClientAuthenticationError("Failed to parse access token from response") from ex
+        else:
+            raise HttpResponseError(
+                message="Failed to exchange Entra token for ACS token",
+                response=response.http_response
+            )
+
+class AsyncTokenExchangeClient:
```

Comment:
Isn't this the same as the TokenExchangeClient and everything in Python is async?